pkg/controller/atlasdeployment/deployment.go

```diff
@@ -1,319 +1,143 @@
-				delete(atlasSpec.RegionsConfig, key)
-			}
-		}
-	}
-}
-
-// DeploymentsEqual compares two Atlas Deployments
-func DeploymentsEqual(log *zap.SugaredLogger, deploymentAtlas mongodbatlas.Cluster, deploymentOperator mongodbatlas.Cluster) bool {
-	deploymentAtlas = removeOutdatedFields(&deploymentAtlas, &deploymentOperator)
-	deploymentOperator = removeOutdatedFields(&deploymentOperator, nil)
 
-	d := cmp.Diff(deploymentAtlas, deploymentOperator, cmpopts.EquateEmpty())
-	if d != "" {
-		log.Debugf("Deployments are different: %s", d)
-	}
-
-	return d == ""
-}
-
-func (r *AtlasDeploymentReconciler) ensureConnectionSecrets(ctx *workflow.Context, project *mdbv1.AtlasProject, name string, connectionStrings *mongodbatlas.ConnectionStrings, deploymentResource *mdbv1.AtlasDeployment) workflow.Result {
-	databaseUsers := mdbv1.AtlasDatabaseUserList{}
-	err := r.Client.List(context.TODO(), &databaseUsers, client.InNamespace(project.Namespace))
-	if err != nil {
-		return workflow.Terminate(workflow.Internal, err.Error())
+		legacyReplicatonSpecs = append(legacyReplicatonSpecs, replicationSpec)
 	}
 
-	secrets := make([]string, 0)
-	for _, dbUser := range databaseUsers.Items {
-		found := false
-		for _, c := range dbUser.Status.Conditions {
-			if c.Type == status.ReadyType && c.Status == v1.ConditionTrue {
-				found = true
-				break
-			}
-		}
-
-		if !found {
-			ctx.Log.Debugw("AtlasDatabaseUser not ready - not creating connection secret", "user.name", dbUser.Name)
-			continue
-		}
-
-		scopes := dbUser.GetScopes(mdbv1.DeploymentScopeType)
-		if len(scopes) != 0 && !stringutil.Contains(scopes, name) {
-			continue
+	for _, legacyResplicationSpec := range legacyReplicatonSpecs {
+		resplicationSpec := &mdbv1.AdvancedReplicationSpec{
+			NumShards:     int(*legacyResplicationSpec.NumShards),
+			ZoneName:      legacyResplicationSpec.ZoneName,
+			RegionConfigs: []*mdbv1.AdvancedRegionConfig{},
 		}
 
-		password, err := dbUser.ReadPassword(r.Client)
-		if err != nil {
-			return workflow.Terminate(workflow.DeploymentConnectionSecretsNotCreated, err.Error())
-		}
-
-		data := connectionsecret.ConnectionData{
-			DBUserName:    dbUser.Spec.Username,
-			ConnURL:       connectionStrings.Standard,
-			SrvConnURL:    connectionStrings.StandardSrv,
-			Password:      password,
-			PvtConnURL:    connectionStrings.Private,
-			PvtSrvConnURL: connectionStrings.PrivateSrv,
-		}
-		connectionsecret.FillPrivateConnStrings(connectionStrings, &data)
-
-		ctx.Log.Debugw("Creating a connection Secret", "data", data)
+		for legacyRegion, legacyRegionConfig := range legacyResplicationSpec.RegionsConfig {
+			regionConfig := mdbv1.AdvancedRegionConfig{
+				AnalyticsSpecs: &mdbv1.Specs{
+					DiskIOPS:      legacy.ProviderSettings.DiskIOPS,
+					EbsVolumeType: legacy.ProviderSettings.VolumeType,
+					InstanceSize:  legacy.ProviderSettings.InstanceSizeName,
+					NodeCount:     toptr.MakePtr(int(*legacyRegionConfig.AnalyticsNodes)),
```

Comment:
Being the NodeCount fields (ElectableNodes, AnalyticsNodes, ReadOnlyNodes) in Legacy Deployments optional and therefore pointers. Once we are not defaulting them, wouldn't be safer to check nil here? (and in the other below)